DPL simulation workflow : Tutorials

sawenzel opened the pull request:
This PR is collecting a series of commits serving as a rough step-by-step tutorial
on how to integrate a new digitization/clusterization processor into the digitizer workflow.
It aims at helping detector experts get a feel for the work necessary.

It demonstrates the use of different DPL constructs such as specification of algorithms in terms
of lambda functions or in terms of the task based approach. Some of the code (especially the writers) will in future be reduced in favour of a generic/configurable device.

sawenzel committed:
DPL simulation workflow : Tutorial Part 1

This is part of a step-by-step tutorial of how to add a new digitizer
to the DPL digitizer workflow. It uses the example of TOF to build the solution
in a series of commits/parts

This is part I which achieves
- definition of the TOF digitization processor
  (in files TOFDigitizerSpec.h and TOFDigitizerSpec.cxx)
- attaching the processor to the SimReader in such a way
  that it can run in parallel to the other digitizers

The system can already be run in a "hello world" mode; The TOF digitizer receives
the collision context from the SimReader and can use the collision times.

A lot of the code is boilerplate and might be reduced in the future by
factoring it out (to services).

sawenzel committed:
DPL simulation workflow : Tutorial Part 2

This is part of a step-by-step tutorial of how to add a new digitizer
to the DPL digitizer workflow. It uses the example of TOF to build the solution
in a series of commits/parts.

This is part 2 which interfaces the actual digitization and transforms hits
to digits.

It adds the following
- add an instance of the o2::tof::Digitizer class which we call
  from the DPL processor function
  (any detector will typically have such a class which implements
   the actual transformation of hits to digits)

- instances of digit and label containers

- read the hits from the input chains

- loop over all collisions and collision parts and pass
  in each case the hits to the digitizer

Note that the logic in the DPL processor function is similar
to the one in typical DigitizerTasks. The main difference is
that we no longer use the FairRootManager to retrieve collisions times
and input.

Note that digits are not yet stored to file. This will be part of
the next step: Forwarding hits to another component (such as one doing IO).

The code here should be seen to describe the rough logic and how
to reuse existing algorithmic components. It is expected that some stuff
will be externalized into common service components (such as retrieval of hits, etc).

chiarazampolli commented:
Ciao,

Thanks Sandro, Francesco and I subscribed to this PR!

Chiara

sawenzel committed:
DPL simulation workflow : Tutorial Part 3

This is part of a step-by-step tutorial of how to add a new digitizer
to the DPL digitizer workflow. It uses the example of TOF to build the solution
in a series of commits/parts.

This is part 3 which provides a consumer processor for TOF digits in terms
of a simple FileSink (TOFDigitWriter).

It adds the following
- add the TOFDigitWriter processor, define its input, and add it to
  the DPL workflow (different consumers such as a clusterizer should look similar)

- writing digits to a configurable file (e.g. tofdigits.root)

- add an output to the TOFDigitizer where digits will be sent

- accumulate all digits in the TOFDigitizer and send/snapshot them

With this part we already have a decently working solution that transforms
hits into digits and stores them into a file, all by using DPL components.

As previously, the code offers many opportunities to factor out stuff into common services.

ktf commented:
This is of course fine, however I was wondering if for the end user exposing the task based API described at:

https://github.com/AliceO2Group/AliceO2/commit/1f67d4ca61eb6de1c3aef8ca64c7bbc64d47cb46

is less confusing than having init which returns the processing callback. Notice that as mentioned by @aphecetche there is no need for a base class either, just something which conforms to having a init and run method.

sawenzel commented:
@ktf: I was not really aware and will take a look into it.

sawenzel commented:
For the record, there is a segfault in `test_Framework_test_SimpleRDataFrameProcessing`.

ktf commented:
Please let me know. I have a few more ideas on how to remove boilerplate I will try to discuss soonish.

I will look into the crash. The mac build should be back in business since I've merged @davidrohr fix in alidist.

shahor02 commented:
Hi @sawenzel 
In your example the digits of the whole TF (interaction records vector) is accumulated then sent as a single vector to the writer. The latter, at least in case of ITS, will need to split it to a vector per detector R/O cycle. This should create quite some memory overhead. The same I see in the TPC digitization code.
Is this caused by the limitation(?) of the DPL that the device can produce only one output per input? 
In principle, would be more economical to flush digits to the tree directly from the DigitizeSpec.
Do you see any problem with avoiding writer component, at least for detectors which can afford this?

Cheers,
 Ruben

ktf commented:
At the moment DPL requires a time identifier to make sure it can dispatch the computation when all the inputs are available. When doing data-processing my understanding is that we will use the timeframe timestamp as such time identifier, because in any case we want to operate on the full timeframe. However there is no restriction in making such a time identifier something else, either larger or smaller quantum, it's enough that whatever produces the incoming data operates on smaller / larger chunks.
In principle it's not even required to have all the data processors use the same clock, as long as there is no mixing between different clocks.

sawenzel commented on Steer/DigitizerWorkflow/src/TOFDigitizerSpec.cxx:
```diff
@@ -0,0 +1,168 @@
+        digitizer->process(&hits, digits.get());
+        // copy digits into accumulator
+        std::copy(digits->begin(), digits->end(), std::back_inserter(*digitsAccum.get()));
+
```
yes but it was not part of the tutorial yet. Will add it next

sawenzel committed:
DPL simulation workflow : Tutorial Part 4

This is part of a step-by-step tutorial of how to add a new digitizer
to the DPL digitizer workflow. It uses the example of TOF to build the solution
in a series of commits/parts.

This is part 4 which provides processors to make and write TOF clusters.

It adds and demonstrates the following

- add the TOFClusterizer processor, define its input, and add it to
  the DPL workflow using the DPL-Task API.

  We demonstrate that the DPL TOFClusterizer used here can be setup
  using a task class implementing 2 methods "init" and "run". This has
  the convenienve of looking more like the existing FairTask mechanism
  used so far and that the implementer does need to use combination of
  lambda functions.

  Note that in comparison to the existing ClustererTask for TOF,
  only few adjustments are needed, which are notably the IO part:
  Instead of getting/sending data via FairRootManager, it is now
  done using the pc.inputs(), pc.outputs().

- write clusters to a configurable file (e.g. tofclusters.root)

For simplicity, Monte Carlo labels are not yet used in this step, which will
be added next.

ktf commented:
You should be able to use arbitrary ids already now. You just need the source to do a the splitting differently by lifting the loop outside the data processing callback. For example given a TTree you could send out one message for one entry, return, and then send a message for the subsequent entry on the subsequent iteration. Wouldn't that work? If the production of different entries is independent and not I/O bound, that could also made to scale quite easily by using `timePipeline`. We can discuss more in front of the whiteboard, if you want.

sawenzel committed:
DPL simulation workflow : Tutorial Part 5

This is part of a step-by-step tutorial of how to add a new digitizer
to the DPL digitizer workflow. It uses the example of TOF to build the solution
in a series of commits/parts.

This is part 5 which add the treatment of MC labels by adding
more input/output specs to the DPL processors.

sawenzel committed:
TOFDigitizer: Add a remaining flush for pileup mode

Also introduce option to disable/enable pileup mode
from DPL workflow
